File: nebras-app/app/src/main/kotlin/com/nebras/mobile/core/service/SearchResultItem.kt
package com.nebras.mobile.core.service

import com.nebras.mobile.core.model.Content
import com.nebras.mobile.feature.home.model.HomeSection

enum class SearchItemType { SECTION, ITEM }

enum class SearchItemSource { FIREBASE }

/** نقل `core/services/search_result_item.dart`. */
data class SearchResultItem private constructor(
    val id: String,
    val title: String,
    val type: SearchItemType,
    val source: SearchItemSource,
    val subtitle: String,
    val imageUrl: String?,
    val score: Double,
    val fuzzy: Boolean,
    val section: HomeSection? = null,
    val content: Content? = null,
) {
    val dedupKey: String
        get() = "${type.name.lowercase()}:${source.name.lowercase()}:$id"

    companion object {
        fun section(
            section: HomeSection,
            source: SearchItemSource,
            score: Double,
            fuzzy: Boolean,
        ): SearchResultItem = SearchResultItem(
            id = section.id,
            title = section.title,
            type = SearchItemType.SECTION,
            source = source,
            subtitle = section.type,
            imageUrl = section.imageUrl,
            score = score,
            fuzzy = fuzzy,
            section = section,
        )

        fun item(
            content: Content,
            source: SearchItemSource,
            score: Double,
            fuzzy: Boolean,
        ): SearchResultItem = SearchResultItem(
            id = content.id,
            title = content.title,
            type = SearchItemType.ITEM,
            source = source,
            subtitle = contentSubtitle(content),
            imageUrl = content.thumbnailUrl.trim().ifEmpty { null },
            score = score,
            fuzzy = fuzzy,
            content = content,
        )

        private fun contentSubtitle(content: Content): String {
            val name = content.sectionName?.trim()
            if (!name.isNullOrEmpty()) return name
            if (content.author.trim().isNotEmpty()) return content.author.trim()
            return content.type.wire
        }
    }
}

File: app-flutter/lib/core/services/search_result_item.dart
import 'package:nebras_mobile_app/features/content/model/content_model.dart';
import 'package:nebras_mobile_app/features/home/data/home_section_model.dart';

enum SearchItemType { section, item }

enum SearchItemSource { firebase }

class SearchResultItem {
  final String id;
  final String title;
  final SearchItemType type;
  final SearchItemSource source;
  final String subtitle;
  final String? imageUrl;
  final double score;
  final bool fuzzy;
  final HomeSection? section;
  final Content? content;

  const SearchResultItem._({
    required this.id,
    required this.title,
    required this.type,
    required this.source,
    required this.subtitle,
    required this.imageUrl,
    required this.score,
    required this.fuzzy,
    this.section,
    this.content,
  });

  factory SearchResultItem.section({
    required HomeSection section,
    required SearchItemSource source,
    required double score,
    required bool fuzzy,
  }) {
    return SearchResultItem._(
      id: section.id,
      title: section.title,
      type: SearchItemType.section,
      source: source,
      subtitle: _sectionSubtitle(section, source),
      imageUrl: section.imageUrl,
      score: score,
      fuzzy: fuzzy,
      section: section,
    );
  }

  factory SearchResultItem.item({
    required Content content,
    required SearchItemSource source,
    required double score,
    required bool fuzzy,
  }) {
    return SearchResultItem._(
      id: content.id,
      title: content.title,
      type: SearchItemType.item,
      source: source,
      subtitle: _contentSubtitle(content),
      imageUrl: content.thumbnailUrl.trim().isEmpty ? null : content.thumbnailUrl,
      score: score,
      fuzzy: fuzzy,
      content: content,
    );
  }

  String get dedupKey => '${type.name}:${source.name}:$id';

  static String _sectionSubtitle(HomeSection section, SearchItemSource source) {
    return section.type;
  }

  static String _contentSubtitle(Content content) {
    final name = content.sectionName?.trim();
    if (name != null && name.isNotEmpty) return name;
    if (content.author.trim().isNotEmpty) return content.author.trim();
    return content.type.name;
  }
}
